### PR TITLE
* add export support for match

### PR DESCRIPTION
**Pull request by hjanuschka**
as i am more and more having the requirement to zip up a signing package for external agency's so they can develop+sign an app for us (wich we deliver and upload to our store account)

i came up with this one.
lets you export the `p12`, `cer` and the `mobileprovision` to a zip file.

usage:
- via standalone command:

``` bash
match export distribution -u helmut@januschka.com -r ssh://git@gitlab.krone.at/mobile/IOS-CERTIFICATES.git -a krone.at-gastro
```
- via fastlane lane

``` ruby
match(
                export: true,
                app_identifier: "krone.at-gastro",
                git_url: "ssh://git@gitlab.krone.at/mobile/IOS-CERTIFICATES.git",
                zip_file: "export.zip",
                username: "helmut@januschka.com"
        )
```

sample output:

```
[10:49:24]: Cloning remote git repo...
[10:49:26]: 🔓  Successfully decrypted certificates repo

+-----------------------+-------------------------------------------------------+
|                        Summary for match export 0.3.0                         |
+-----------------------+-------------------------------------------------------+
| username              | helmut@januschka.com                                  |
| git_url               | ssh://git@gitlab.krone.at/mobile/IOS-CERTIFICATES.git |
| app_identifier        | krone.at-gastro                                       |
| force                 | true                                                  |
| type                  | appstore                                              |
| zip_file              | export.zip                                            |
| keychain_name         | login.keychain                                        |
| readonly              | false                                                 |
| verbose               | false                                                 |
| export                | false                                                 |
| shallow_clone         | false                                                 |
| force_for_new_devices | false                                                 |
+-----------------------+-------------------------------------------------------+

[10:49:26]: Fetching certificates and profiles...
+-----------------------------------------------------------------------------------------------------------+
|                                Certificates that are going to be exported                                 |
+-----------------------------------------------------------------------------------------------------------+
| Path                                                                                                      |
+-----------------------------------------------------------------------------------------------------------+
| /var/folders/dx/z94090d97ydbnpk21m4clpjjqqx4h1/T/d20160316-11783-7jq5f6/certs/distribution/J2BB497FT7.cer |
+-----------------------------------------------------------------------------------------------------------+

+-----------------------------------------------------------------------------------------------------------+
|                                    Keys that are going to be exported                                     |
+-----------------------------------------------------------------------------------------------------------+
| Path                                                                                                      |
+-----------------------------------------------------------------------------------------------------------+
| /var/folders/dx/z94090d97ydbnpk21m4clpjjqqx4h1/T/d20160316-11783-7jq5f6/certs/distribution/J2BB497FT7.p12 |
+-----------------------------------------------------------------------------------------------------------+

+------------------------------------------------------------------------------------------------------------------------------------+
|                                        Provisioning Profiles that are going to be exported                                         |
+------------------------------------------------------------------------------------------------------------------------------------+
| Path                                                                                                                               |
+------------------------------------------------------------------------------------------------------------------------------------+
| /var/folders/dx/z94090d97ydbnpk21m4clpjjqqx4h1/T/d20160316-11783-7jq5f6/profiles/appstore/AppStore_krone.at-gastro.mobileprovision |
+------------------------------------------------------------------------------------------------------------------------------------+

[10:49:32]: Exported to: export.zip
[10:49:32]: 🔒  Successfully encrypted certificates repo
```

inside the zip it looks like:

```
Archive:  export.zip
  Length     Date   Time    Name
 --------    ----   ----    ----
     1473  03-16-16 10:49   certs/distribution/J2BB497FT7.cer
     1675  03-16-16 10:49   certs/distribution/J2BB497FT7.p12
     7830  03-16-16 10:49   profiles/appstore/AppStore_krone.at-gastro.mobileprovision
 --------                   -------
    10978                   3 files
```

as a note: maybe i do another PR to get `verify_gem` into FastlaneCore

### CODE REVIEW

File: match/bin/match [hunk 1]
```diff
@@ -82,6 +82,25 @@ class MatchApplication
       end
     end
 
+    command :export do |comm|
+      comm.syntax = "match export"
+      comm.description = "Exports Provisiong Profile + Certificate + P12"
+      comm.action do |args, options|
+        # FastlaneCore::UI.user_error!("Please run `match export [type], allowed values: distribution and development")
```
Comment (KrauseFx) [inline match/bin/match]: Mh?

File: match/lib/match/export.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    attr_accessor :certs
+    attr_accessor :keys
+    attr_accessor :profiles
+    attr_accessor :files
+
+    def run(params, type: nil)
+      self.params = params
+      self.type = type
+      params[:workspace] = GitHelper.clone(params[:git_url], params[:shallow_clone])
+      FastlaneCore::PrintTable.print_values(config: params,
```
Comment (KrauseFx) [inline match/lib/match/export.rb]: This file seems like a lot of duplicate code from what we already have in `match`, right?

**Comment by KrauseFx** [comment]
Thanks for the PR @hjanuschka. I wonder if we can find an easier way to solve the same problem, maybe by just decrypting the repo and giving the user access to it?

File: match/bin/match (hunk 1 above)
Comment (hjanuschka) [inline match/bin/match]: thx, i fucked this up!  - stupid ruby noob!

File: match/lib/match/export.rb (hunk 2 above)
Comment (hjanuschka) [inline match/lib/match/export.rb]: yeah i took the `nuke.rb` as a starting point.

**Comment by hjanuschka** [comment]
thx for takeing the time to  review, fixed the feedback from the first comment.

about the code-duplication, i took `nuke.rb` as a starting point - well i am pretty ruby noob - so if you have suggestions on how to do it a cleaner way, i am open for feedback.


yeah decrypting the whole repo, and leaving it arround, is another solution, but in my case i have like 20 apps in the repo, and just want one to be exported, in a zip file.


maybe another command `match decrypt` would be better for your suggested-case?


@KrauseFx

**Comment by KrauseFx** [comment]
Thanks for updating the PR :+1: I'd really much rather prefer not to have too much duplicate code, since it will require a lot of extra work in case we change the directory structure. I'd be okay with a `match decrypt` command that just clones the repo and decrypts it :rocket:

**Comment by hjanuschka** [comment]
see https://github.com/fastlane/fastlane/pull/3831 for the decrypt part.


for this PR i have following ideas, please let me if one of this is ok for you, then i ll update this PR:

  * make an action `match_decrypt` -> wich decrypts the repo (so we can do it via fastfile, and set shared variable to be used in later actions).
  * make an action `match_export` -> wich extracts profiles based on app-id, and creates a zip (like in this PR)
  * centralize  the file finding stuff (where you grep for p12,cer and so on) to be used by `match_export` and `nuke` - to minimize code duplication (something like `repo_file_helper.rb`)


i would love to keep this one open, if the above ideas are ok for you - and handle the "decrypt only" in the other PR.

@KrauseFx

**Comment by KrauseFx** [comment]
Thanks for the PR @hjanuschka, I'm not 100% sold yet. We already have the export command now, that just encrypts the whole repo. I think we should just stick with that, what do you think?

**Pull request closed without merging**